crates/wasm-mutate/src/mutators/peephole/rules.rs

```diff
@@ -19,656 +19,687 @@ impl PeepholeMutator {
+    fn add_bidirectional_rewrite(
         &self,
-        vari: &'static str,
-        t: PrimitiveTypeInfo,
-    ) -> impl Fn(&mut EG, Id, &Subst) -> bool {
-        move |egraph: &mut EG, _, subst| {
-            let var = vari.parse();
-            match var {
-                Ok(var) => {
-                    let eclass = &egraph[subst[var]];
-                    match &eclass.data {
-                        Some(d) => d.tpe == t,
-                        None => false,
-                    }
-                }
-                Err(_) => false,
-            }
-        }
+        rules: &mut Vec<Rewrite<Lang, PeepholeMutationAnalysis>>,
+        name: &str,
+        lhs: &str,
+        rhs: &str,
+        cond: &[Condition],
+    ) {
+        self.add_rewrite(rules, name, lhs, rhs, cond.clone());
+        self.add_rewrite(rules, &format!("{name}-rev"), rhs, lhs, cond);
     }
 
-    /// Check whether the given variable is a constant.
-    fn is_const(&self, vari: &'static str) -> impl Fn(&mut EG, Id, &Subst) -> bool {
-        move |egraph: &mut EG, _, subst| {
-            let var = vari.parse();
-            match var {
-                Ok(var) => {
-                    let eclass = &egraph[subst[var]];
-                    if eclass.nodes.len() == 1 {
-                        let node = &eclass.nodes[0];
-                        match node {
-                            Lang::I32(_) => true,
-                            Lang::I64(_) => true,
-                            Lang::F32(_) => true,
-                            Lang::F64(_) => true,
-                            _ => false,
-                        }
-                    } else {
-                        false
+    fn add_rewrite(
+        &self,
+        rules: &mut Vec<Rewrite<Lang, PeepholeMutationAnalysis>>,
+        name: &str,
+        lhs: &str,
+        rhs: &str,
+        cond: &[Condition],
+    ) {
+        let name = name.to_string();
+        let long_name = format!("{lhs} => {rhs}");
+        let lhs = lhs.parse::<egg::Pattern<_>>().unwrap();
+        let rhs = rhs.parse::<egg::Pattern<_>>().unwrap();
+        rules.push(match cond.len() {
+            0 => Rewrite::new(name, long_name, lhs, rhs).unwrap(),
+            1 => Rewrite::new(
+                name,
+                long_name,
+                lhs,
+                egg::ConditionalApplier {
+                    condition: cond[0].clone(),
+                    applier: rhs,
+                },
+            )
+            .unwrap(),
+            2 => Rewrite::new(
+                name,
+                long_name,
+                lhs,
+                egg::ConditionalApplier {
+                    condition: cond[0].clone(),
+                    applier: egg::ConditionalApplier {
+                        condition: cond[1].clone(),
+                        applier: rhs,
+                    },
+                },
+            )
+            .unwrap(),
+            _ => unimplemented!(),
+        });
+    }
```

Comment:
Rather than forking `egg` code, what if we just split the rule set into a bunch of different functions based on how they are grouped together? Not required, but it would be nice to avoid copying code out if we can...

Reply:
While possible I'm honestly not too gung-ho about redoing this in the sense that it isn't really changing that much and we don't have a ton of eyes on it at this time. We're pretty far behind `egg` at this point now as well as well so sticking close to `egg` is sort of a sailed ship at this point. 

All that being said though I think that this variant is concretely less code since it's moving all the rule parsing/compliation into one function instead of having it inlined and monomorphized in lots of locations. The current `egg` API isn't too different from this so updating in theory wouldn't be too bad.

I suppose another way to put all that: how strongly do you feel that this should continue to use the macro as-is from `egg`?

Reply:
Not strong enough to really argue the point. Let's merge!